### PR TITLE
Add New Cloud Provider: Samsung Cloud Platform

### PR DESCRIPTION
**Pull request by JGoo1**
The following PR is about the addition of Samsung Cloud Platform (SCP) to Sky Pilot as a new cloud provider.
Smoke tests have been completed, and formatting for the added code has been done using 'format.sh'.

The smoke test results with the "--scp" option are as follows:

================================================= FAILURES =================================================
___________________________________________ test_azure_disk_tier ___________________________________________
[gw6] linux -- Python 3.7.16 /home/jgoo1/anaconda3/envs/sky/bin/python
tests/test_smoke.py:1994: in test_azure_disk_tier
    run_one_test(test)
tests/test_smoke.py:181: in run_one_test
    raise Exception(f'test failed: less {log_file.name}')
E   Exception: test failed: less /tmp/azure-disk-tier-xbnut0vc.log
========================================= short test summary info ==========================================
FAILED tests/test_smoke.py::test_azure_disk_tier - Exception: test failed: less /tmp/azure-disk-tier-xbnut0vc.log
1 failed, 34 passed, 68 skipped, 960 warnings in 2069.05s (0:34:29)

Out of a total of 35 test cases, 1 failed, and no further tests were performed because the failing cases were not related to SCP function.

No other unit tests were been performed except for smoke tests.

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@JGoo1 @sukhoonjung Thank you for submitting the PR! Before proceeding with the review, please ensure that the code formatting adheres to our guidelines, as it seems that the linter check was not successful.

**Review by WoosukKwon — COMMENTED** [review]
Thanks for the great work! I've checked that the core features (launch, stop, autostop, down) correctly work on SCP. I only have minor concerns:

0. Please make sure the PR satisfies our formatting guidelines.
1. It seems VM provisioning is a bit slow (5 -- 7 minutes) in SCP.
2. It seems adding SCP slows down SkyPilot. 
```
# Current master
$ time echo 'n' | sky gpunode
real    0m1.369s
user    0m0.813s
sys     0m0.111s

# This PR
$ time echo 'n' | sky gpunode
real    0m1.955s
user    0m0.870s
sys     0m0.104s
```
I guess some SCP http requests are invoked every time SkyPilot is loaded.

3. I got a VPC error when launching a CPU VM. Is this expected? Please check it out.
```
$sky cpunode --cpus 16+ -c woosuk-cpu --cloud scp
I 05-15 22:53:59 optimizer.py:636] == Optimizer ==
I 05-15 22:53:59 optimizer.py:647] Target: minimizing cost
I 05-15 22:53:59 optimizer.py:659] Estimated cost: $0.6 / hour
I 05-15 22:53:59 optimizer.py:659] 
I 05-15 22:53:59 optimizer.py:732] Considered resources (1 node):
I 05-15 22:53:59 optimizer.py:781] -------------------------------------------------------------------------------------------------
I 05-15 22:53:59 optimizer.py:781]  CLOUD   INSTANCE   vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE             COST ($)   CHOSEN   
I 05-15 22:53:59 optimizer.py:781] -------------------------------------------------------------------------------------------------
I 05-15 22:53:59 optimizer.py:781]  SCP     s1v16m32   16      32        -              KOREA-EAST-1-SCP-B001   0.56          ✔     
I 05-15 22:53:59 optimizer.py:781] -------------------------------------------------------------------------------------------------
I 05-15 22:53:59 optimizer.py:781] 
Launching a new cluster 'woosuk-cpu'. Proceed? [Y/n]: y
Setting up interactive node woosuk-cpu...
I 05-15 22:53:59 cloud_vm_ray_backend.py:3485] Creating a new cluster: "woosuk-cpu" [1x SCP(s1v16m32)].
I 05-15 22:53:59 cloud_vm_ray_backend.py:3485] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
I 05-15 22:54:00 cloud_vm_ray_backend.py:1241] To view detailed progress: tail -n100 -f /Users/woosuk/sky_logs/sky-2023-05-15-22-53-58-386875/provision.log
I 05-15 22:54:00 cloud_vm_ray_backend.py:1565] Launching on SCP KOREA-EAST-1-SCP-B001
W 05-15 22:54:04 cloud_vm_ray_backend.py:872] Got error(s) in KOREA-EAST-1-SCP-B001:
W 05-15 22:54:04 cloud_vm_ray_backend.py:874]   sky.skylet.providers.scp.node_provider.SCPError: This region/zone does not have available VPCS.
```

File: docs/source/getting-started/installation.rst
Comment (JGoo1) [reply docs/source/getting-started/installation.rst:161]: You're right. I'll fix it. Thanks :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:16]: fixed it. :)

File: sky/skylet/providers/scp/node_provider.py
Comment (dreamersnme) [reply sky/skylet/providers/scp/node_provider.py:411]: fixed it. :)

File: sky/skylet/providers/scp/scp_utils.py
Comment (dreamersnme) [reply sky/skylet/providers/scp/scp_utils.py:10]: fixed it. :)

File: sky/skylet/providers/scp/scp_utils.py
Comment (dreamersnme) [reply sky/skylet/providers/scp/scp_utils.py:18]: fixed it. :)

File: sky/skylet/providers/scp/scp_utils.py
Comment (dreamersnme) [reply sky/skylet/providers/scp/scp_utils.py:5]: fixed it. :)

File: sky/clouds/service_catalog/scp_catalog.py
Comment (dreamersnme) [reply sky/clouds/service_catalog/scp_catalog.py:18]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:30]: fixed it :)

File: sky/backends/backend_utils.py
Comment (JGoo1) [reply sky/backends/backend_utils.py:1788]: No, it is not related to the cost. 
It simply means that 'Stopping' or 'Terminating' refers to the state where the process is still in progress before it is 'Stopped' or 'Terminated'

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:249]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:342]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:309]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:299]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:321]: fixed it. :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:129]: fixed it :)

File: sky/clouds/scp.py
Comment (dreamersnme) [reply sky/clouds/scp.py:136]: removed. :)

File: sky/templates/aws-ray.yml.j2
Comment (dreamersnme) [reply sky/templates/aws-ray.yml.j2:185]: fixed it.

File: docs/source/getting-started/installation.rst
Comment (dreamersnme) [reply docs/source/getting-started/installation.rst]: fixed it :)

**Comment by dreamersnme** [comment]
> * Please make sure the PR satisfies our formatting guidelines.
> * It seems VM provisioning is a bit slow (5 -- 7 minutes) in SCP.
> * It seems adding SCP slows down SkyPilot.
> * I got a VPC error when launching a CPU VM. Is this expected? Please check it out.

**Comment by dreamersnme** [comment]
> 0. Please make sure the PR satisfies our formatting guidelines.
We have executed formatting.sh again, and changed some codes. please check it. 

> 1. It seems VM provisioning is a bit slow (5 -- 7 minutes) in SCP.
It is right. SCP has several networking/security features managed during VM provisioning. 
Those make the provisioning a bit slow. 

> 2. It seems adding SCP slows down SkyPilot.
> I guess some SCP http requests are invoked every time SkyPilot is loaded.
You are right that, when SCP is Enabled, the catalog service invokes SCP HTTP requests.  
An SCP user is allowed to access not all but only designated service zones(=regions) configured in the SCP project.   
So it is necessity to obtain the available service zones, via HTTP requests during the catalog service.
We hope this feature won't be too much of a problem for you, since it happens only SCP is used.

> 3. I got a VPC error when launching a CPU VM. Is this expected? Please check it out.
It was expected. We deigned the provider to search only available VPCs, which are open to public. 
Automatically creating a public VPC via Skypilot may violate SCP security regulations, so I leave it as a manual step in the SCP project configuration.

**Comment by dreamersnme** [comment]
> 0. Please make sure the PR satisfies our formatting guidelines.
We have executed formatting.sh again, and changed some codes. please check it. 

> 1. It seems VM provisioning is a bit slow (5 -- 7 minutes) in SCP.
It is right. SCP has several networking/security features managed during VM provisioning. 
Those make the provisioning a bit slow. 

> 2. It seems adding SCP slows down SkyPilot.
> I guess some SCP http requests are invoked every time SkyPilot is loaded.
It is right that, when SCP is Enabled, the catalog service invokes an SCP HTTP request.  
An SCP user is allowed to access not all but only designated service zones(=regions) configured in the SCP project.   
So it is necessity to obtain the available service zones, via HTTP requests during the catalog service.
We hope this feature won't be too much of a problem for you, since it happens only SCP is used.
Still.. it is weird that one HTTP request with small response data makes it 0.6 sec. slower...
If you have another opinion on that, please let us know. 


> 3. I got a VPC error when launching a CPU VM. Is this expected? Please check it out.
It was expected. We deigned the provider to search only available VPCs, which are open to public. 
Automatically creating a public VPC via Skypilot may violate SCP security regulations, so I leave it as a manual step in the SCP project configuration.

**Comment by WoosukKwon** [comment]
@dreamersnme @JGoo1 It seems the PR still doesn't pass our formatting requirements. Could you please re-run `format.sh` and resolve the formatting issues?

**Comment by dreamersnme** [comment]
@WoosukKwon , we found there was an mistake on mypy configuration, sorry.   For some reason, mypy in format.sh was not working properly  in our environments, so we ran it independently. I hope  it wont cause unexpected inspection results. Thanks.

**Comment by dreamersnme** [comment]
@WoosukKwon , we found there was an mistake on mypy configuration, sorry.  We updated the codes, please check them again.
For some reason, mypy in format.sh was not working properly in our environments, so we ran it independently. I hope it wont cause unexpected inspection results. Thanks.

**Comment by dreamersnme** [comment]
@WoosukKwon , we found there was an mistake on mypy configuration, sorry. We updated the codes, please check them again.
For some reason, mypy in format.sh was not working properly in our environments, so we ran it independently. I hope it wont cause unexpected inspection results. Thanks.

File: sky/clouds/service_catalog/scp_catalog.py
Comment (WoosukKwon) [inline sky/clouds/service_catalog/scp_catalog.py:29]: > 2. It seems adding SCP slows down SkyPilot.
> I guess some SCP http requests are invoked every time SkyPilot is loaded.
> It is right that, when SCP is Enabled, the catalog service invokes an SCP HTTP request. (checked it takes 0.2~0.5 secs)
> An SCP user is allowed to access not all but only designated service zones(=regions) configured in the SCP project.
> So it is necessity to obtain the available service zones, via HTTP requests during the catalog service.
> We hope this feature won't be too much of a problem for you, since it happens only SCP is used.

@dreamersnme Can we avoid the slowdown by caching the data locally and update them less frequently? For example, here the `list_service_zone_names` method is called every time the scp module is loaded. Why don't we cache this data locally?

**Comment by WoosukKwon** [comment]
Another concern with the HTTP calls is that the same APIs are being called more quite redundantly.
For example, 9 HTTP calls are made to terminate a VM.
```
$ sky down sky-cpunode-woosuk
Terminating 1 cluster: sky-cpunode-woosuk. Proceed? [Y/n]: y
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers/INSTANCE-5Rft-svpttnOjSZualOfwg/nics
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
Terminating cluster sky-cpunode-woosuk...done.
Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
```
Can we optimize this?

**Comment by dreamersnme** [comment]
> Another concern with the HTTP calls is that the same API is being called quite redundantly. For example, 9 HTTP calls are made to terminate a VM.
> 
> ```
> $ sky down sky-cpunode-woosuk
> Terminating 1 cluster: sky-cpunode-woosuk. Proceed? [Y/n]: y
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers/INSTANCE-5Rft-svpttnOjSZualOfwg/nics
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/virtual-server/v2/virtual-servers
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--get https://openapi.samsungsdscloud.com/security-group/v2/security-groups
> Terminating cluster sky-cpunode-woosuk...done.
> Terminating 1 cluster ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
> ```
> 
> Can we optimize this?

These are made to check status by polling whether a request has been completed or not. 
We could increase the polling interval to reduce the number of requests, but using a long interval may cause lack of real-time recognition.

**Comment by WoosukKwon** [comment]
> These are made to check status by polling whether a request has been completed or not.
> Some resource creation/termination takes time, so needs to be checked by polling before the next step.
> We could increase the polling interval to reduce the number of requests, but using a long interval may cause lack of real-time recognition.

I see. But I believe at least the SCP APIs should be never called by the commands that do not really involve SCP VMs (e.g., `sky show-cpus`, `sky gpunode --gpus T4`).

**Comment by WoosukKwon** [comment]
@dreamersnme Besides, please make sure the PR passes all the unit tests.

**Comment by dreamersnme** [comment]
@WoosukKwon Rest calls in the catalog service has been removed. please check it.

**Comment by WoosukKwon** [comment]
@dreamersnme I've rerun the CI. Please check the error messages in the failed test.

**Comment by dreamersnme** [comment]
@WoosukKwon We fixed bugs and, on a local PC, passed tests which had failed/canceled in last CI tests.  
please run the tests again.  Thanks. 

tests/test_global_user_state.py
tests/test_list_accelerators.py
tests/test_optimizer_dryruns.py
tests/test_spot.py

**Comment by WoosukKwon** [comment]
@dreamersnme @JGoo1 The PR now passes all the tests 😄. It also looks good to me. @concretevitamin Can you take a look?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3174]: nit: it would be enough to have the exception in one of the stdout/stderr.
```suggestion
                stdout = ''
                stderr = str(e)
```

File: sky/clouds/scp.py
Comment (Michaelvll) [inline sky/clouds/scp.py:326]: nit: use parentheses instead of `\` for the code style.
```suggestion
        if (resources.disk_size and
               (resources.disk_size < _SCP_MIN_DISK_SIZE_GB or
                resources.disk_size > _SCP_MAX_DISK_SIZE_GB)):
```

File: sky/templates/scp-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/scp-ray.yml.j2:99]: We recently updated the ports for the ray cluster. Please follow the following lines to update the commands here https://github.com/skypilot-org/skypilot/blob/8a48e9be27768ddd36d7187628bb98edbcc26051/sky/templates/aws-ray.yml.j2#L205-L213

File: sky/templates/scp-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/scp-ray.yml.j2:104]: ditto. Please update the ports

**Review by Michaelvll — COMMENTED** [review]
Thanks for the PR @dreamersnme @JGoo1 and @WoosukKwon's review! It is very exciting to see the support of SDS. I went through most of the files except for the `node_providers`.
Due to the timeline, we can merge the PR first once the master branch is merged, comments are resolved and the tests pass. We can return to the node_providers later to see if we have anything left out. : )

**Comment by dreamersnme** [comment]
@Michaelvll Thank you for spending your valuable time on this review! We updated the codes according to your feedback. The modification has been tested and confirmed that it worked properly,  along with your recent updates made in the master branch. 
 Thank you!.

File: sky/templates/scp-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/scp-ray.yml.j2]: Do we want to still keep these lines? It stops the unattended upgrade for apt, which can hold the lock for the user trying `sudo apt install`. We need to confirm whether the unattended upgrade is enabled in the image we are using.
We don’t include this in the `aws-ray.yaml` because we are using the cloud-init in the following place to disable the unattended upgrade.
https://github.com/skypilot-org/skypilot/blob/88c7a9e277171aaa6ba5086abce52b0c4e6f421f/sky/templates/aws-ray.yml.j2#L78-L87

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @dreamersnme! It mostly looks good to me! Just left 2 final comments, and we should be able to merge it.

File: sky/clouds/service_catalog/scp_catalog.py
Comment (Michaelvll) [inline sky/clouds/service_catalog/scp_catalog.py:127]:
```suggestion
        gpus_only: bool,
        name_filter: Optional[str],
        region_filter: Optional[str],
        quantity_filter: Optional[int],
        case_sensitive: bool = True
```

File: sky/clouds/service_catalog/scp_catalog.py
Comment (Michaelvll) [inline sky/clouds/service_catalog/scp_catalog.py:131]:
```suggestion
                                         region_filter, quantity_filter, case_sensitive)
```

**Comment by WoosukKwon** [comment]
I will merge this PR first and then submit a PR to fix the CI issues.

**Pull request merged**
